---
title: "A Global Transcriptional Atlas of the Effect of Sleep Loss in the Mouse Cortex"
font_size: 12pt
output: pdf_document
toc: TRUE
date: ""
bibliography: biblio.bib
link-ciations: TRUE
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	fig.align = "center",
	fig.show = "asis",
	message = FALSE,
	warning = FALSE,
	cache = FALSE,
	results = "markup",
	size = "small",
	tidy = TRUE,
	tidy.opts = list(width.cutoff = 60)
)
# cache=TRUE,cache.lazy = FALSE
options(width=65)
library(BiocStyle)
library(knitr)

```

# Differential Transcript and Differential Gene Expression using Fishpond with RUVs

## Description

Sleep deprivation (SD) has negative effects on brain function. Sleep problems are prevalent in neurodevelopmental, neurodegenerative and psychiatric disorders. Thus, understanding the molecular consequences of SD is of fundamental importance in neuroscience. In this study we present the first simultaneous bulk and single single-nuclear (sn)RNA sequencing characterization of the effects of SD in the prefrontal cortex (PFC). We show that SD predominantly affects glutamatergic neurons, specifically in layers 4 and 5, and produces isoform switching of thousands of transcripts. At both the global and cell-type specific level, SD has a large repressive effect on transcription, down-regulating thousands of genes and transcripts; underscoring the importance of accounting for the effects of sleep loss in transcriptome studies of brain function. As a resource we provide extensive characterization of cell-types, genes, transcripts and pathways affected by SD; as well as tutorials for data analysis.

This code was adapted from the tutorial here: <https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html#Quick_start>

## Data Availability

"Raw" sequencing data for this analysis is stored in GEO under accession number [GSE211088](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE211088). Additionally, the bulk files were previously published [@ingiosi2019] and can be accessed in GEO under accession number [GSE113754](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE113754).

The code used in this analysis has been deposited into Github, and can be accessed via the following link: **link to github folder where this code is**

Prior to this analysis, files were quantified using *Salmon* [@patro2017] and imported into *tximeta* [@love2020] to generate a *SummarizedExperiment* object.

## Data Import

Establish the working directory using the code below. The working directory can be thought of as the default location where files can be accessed from or saved to.

```{r}
setwd("~/Dropbox/Sleep_RNAseq_Splicing/BulkRNAseq_SD/Final_Tximeta")
dir <- setwd("~/Dropbox/Sleep_RNAseq_Splicing/BulkRNAseq_SD/Final_Tximeta")
```

Import the colData, which contains the sample names and condition (the condition is necessary for Fishpond). In short, the colData is a file that contains descriptions of the samples, and is created during *tximeta* [@love2020]. You can see what the columns and first few rows of the file look like using the *head()* function:

```{r}
coldata <- read.table(file.path(dir, "coldata.txt"))
head(coldata)
```

Add a path to locate the files:

```{r}
coldata$files <- file.path(dir, "data/Salmon_Quants", coldata$names, "quant.sf")
```

Before continuing with the analysis, make sure that all files in your colData exist in this location:

```{r}
all(file.exists(coldata$files))
```

Load *SummarizedExperiment* and *tximeta* packages:

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment)) # Version 1.28.0
suppressPackageStartupMessages(library(tximeta)) # Version 1.16.1
```

Load the quantification data with *tximeta*:

```{r}
se <- tximeta(coldata) 
```

Ensure assays are loaded by viewing *assayNames()*. Assays are experimental results, and are necessary for *fishpond*. Call *assay()* to retrieve experimental data.

```{r}
assayNames(se)
```

View rownames (transcript IDs) using the *head()* function:

```{r}
head(rownames(se))
```

# Differential Transcript Expression Analysis Starts Here

'se' will be referred to as 'y' for the remainder of the analysis.

```{r}
y <- se
```

I will do a two group comparison between wild-type animals. Those that were left undisturbed in their home cage (HC) for 5 hours and those who were sleep sleep deprived (SD) via gentle handling for 5 hours.

The following analysis will compare the second group to the first group listed in *levels*, in this case, SD compared to HC controls.

```{r}
y <- y[,y$condition %in% c("WTHC5", "WTSD5")]
y$condition <- factor(y$condition, levels=c("WTHC5", "WTSD5"))
```

## Load Fishpond, and Proceed with Differential Expression Analysis

Load the *fishpond* package to proceed with differential expression analysis:

```{r}
suppressPackageStartupMessages(library(fishpond)) # Version 2.4.1
```

Scale inferential replicates to the mean sequencing depth. It is important to note that only the inferential replicates are scaled during this process, not the counts, which will be relevant in later steps.

```{r}
y <- scaleInfReps(y)
```

Before filtering, we have 140,992 transcripts:

```{r}
dim(y)
```

Filter here, keeping a minimum count of 10 transcripts across a minimum of 3 replicates. In this step, we are removing most of the rows with zero to very low counts before proceeding with the analysis.

```{r}
y <- labelKeep(y, minCount = 10, minN = 3) 
```

After filtering, we have 54,030 transcripts:

```{r}
y <- y[mcols(y)$keep,]
dim(y)
```

## Determining the Sources of Unwanted Variation with RUVs

Assemble a matrix that contains the groups. If you have uneven groups, "fill" the extra spaces in the matrix with "-1" values.

```{r}
groups <- matrix(data = c(1:5, 6:10), nrow = 2, byrow = TRUE)
```

Check to make sure 'groups' appears as expected.

```{r}
groups
```

Assign the names of the variables to match the column names of your data. This is important for *RColorBrewer*, which is used to designate colors for figures later on in this analysis:

```{r}
x <- as.factor(rep(c("WTHC5", "WTSD5"), c(5,5)) )
x
```

Shorten sample names for simplicity while making the matrix:

```{r}
names(x) <- c("HC5_1", "HC5_2", "HC5_3", "HC5_4", "HC5_5", "SD5_1", "SD5_2", 
              "SD5_3", "SD5_4", "SD5_5")
```

See what type of data class 'x' is using the code below:

```{r}
data.class(x)
```

Turn 'x' into a matrix:

```{r}
as.matrix(x)
```

Load *RColorBrewer* here:

```{r}
suppressPackageStartupMessages(library(RColorBrewer)) # Version 1.1-3
```

This palette will be used for PCA and RLE plots for both transcript and gene expression analysis. HC controls will be light grey and SD samples will be dark grey. To view other palette options, use *brewer.pal.info*:

```{r}
colors <- brewer.pal(9, "Greys")[c(6, 7)]
colLib <- colors[x]
HC_color  <- brewer.pal(9, "Greys")[6] 
SD_color <- brewer.pal(9, "Greys")[7] 
```

Remove the version #s for each transcript. It is helpful to note that the version #s correspond to the release version, they are not different transcripts of the same gene:

```{r}
rownames(y) <- lapply(rownames(y),  sub, pattern = "\\.\\d+$", replacement = "")
data.class(rownames(y))
```

Extract the counts from the *SummarizedExperiment* object for normalization. Note that we will estimate the factors of unwanted variation on the counts, and then we will apply that across the inferential replicates further downstream:

```{r}
counts <- as.matrix((assays(y)[["counts"]]))
data.class(counts)
```

Upper Quartile (UQ) normalization is implemented here with the *EDASeq* package, and accounts for variation in sequencing depth:

As the counts were not scaled (only the inferential replicates), we need to account for sequencing depth prior to proceeding with *RUVs* normalization, which allows us to obtain a the estimated factors of unwanted variation (*s\$W*):

```{r}
suppressPackageStartupMessages(library("EDASeq")) # Version 2.32.0
uq <- betweenLaneNormalization(counts, which = "upper")
dim(uq)
```

Shorten the sample names (*colnames*) for figures:

```{r}
colnames(uq)
colnames(uq) <- c("HC5_1", "HC5_2", "HC5_3", "HC5_4", "HC5_5", "SD5_1", "SD5_2", 
                  "SD5_3", "SD5_4", "SD5_5")
```

Here we plot two quality control plots (RLE and PCA) following UQ normalization. Note that triangles are sleep deprived animals and circles are home cage control animals that were collected at the same circadian time:

```{r}
plotRLE(uq, col= colLib, outline = FALSE, las = 3, ylim = c(-1, 1), 
        ylab = "Relative Log Expression", cex.axis = 1.3, cex.lab = 1.3)
plotPCA(uq, labels = FALSE, col = colLib, cex = 1.5, cex.axis = 1.3, cex.lab = 1.3, 
        xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75), pch = rep(c(19,17), times = c(5,5)))
```

Load *RUVSeq* package:

```{r}
suppressPackageStartupMessages(library(RUVSeq)) # Version 1.32.0
```

*RUVseq*: Here we will estimate a matrix that contains estimated factors of unwanted factors after UQ normalization.

*RUVs* uses technical replicates or negative controls samples. In this instance, as we do not have negative controls for transcripts, we will use the rownames of the matrix y after filtering (all expressed transcripts). For additional information regarding *RUVs*, see [@risso2014].

We have set the numbers of factors of unwanted variation (*k*) = 4, which optimizes the positive control recovery (at the gene level) without removing signal. When one is beginning analysis, it is recommended to test a variety of *k*s, and compare the results to one another.

```{r}
s <- RUVs(x = uq, cIdx = rownames(y), scIdx = groups, k = 4)
```

Plot the RLE and PCA plots (quality control figures) again following *RUVs* normalization. After proper normalization, the largest source of variation in the dataset is the treatment as visualized on the PCA plot (separated along PC1).

```{r}
plotRLE(s$normalizedCounts, col= colLib, outline = FALSE, las = 3, 
        ylim = c(-1, 1), ylab = "Relative Log Expression", cex.axis = 1.2, 
        cex.lab = 1.2)
plotPCA(s$normalizedCounts, labels = FALSE, col = colLib, cex = 1.3, 
        cex.axis = 1.2, cex.lab = 1.2, xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75), 
        pch = rep(c(19,17), times = c(5,5)))
```

Downstream we will use *s\$W* (from *RUVs*) as the estimation of batch factors.

'infRepIdx' lists the inferential replicates (30):

```{r}
infRepIdx <- grep("infRep",assayNames(y),value=TRUE)
```

Save the number of inferential replicates as 'nreps':

```{r}
nreps <- length(infRepIdx)
nreps
```

Account for continuous variables with *removeBatchEffect* from *limma*. Our samples do not fall into discrete clusters so we will use the following approach recommended by the *fishpond* developers.

In short, this is done by directly scaling the estimated counts across inferential replicates.

First inferential replicates are logged as *limma* requires log-expression values for a series of samples. This is done with the assay function from the *SummarizedExperiment* package.

First, we load the *limma* package:

```{r}
suppressPackageStartupMessages(library(limma)) # Version 3.54.2
```

```{r}
model_matrix <- model.matrix(~condition, colData(y))
pc <- .1 # This is added to avoid negative InfReps
for (k in seq_len(nreps)) {
  logInfRep <- log(assay(y, infRepIdx[k]) + pc)
  logInfRep <- limma::removeBatchEffect(
    logInfRep,
    covariates=s$W,
    design=model_matrix)
  assay(y, infRepIdx[k]) <- exp(logInfRep)
}
```

## Run Swish and Plot Additional Figures

The Swish method is described in [@zhu2019]. Note that the *set.seed* function allows for the reproducibility of exact results in the future.

```{r}
set.seed(1)
y <- swish(y, x="condition")
```

View a table of differently expressed transcripts with a qvalue \< 0.05:

```{r}
table(mcols(y)$qvalue < .05)
```

You can also view a table that shows which transcripts are up regulated (1) in response to sleep deprivation and which are downregulated (-1):

```{r}
with(mcols(y),
     table(sig=qvalue < .05, sign.lfc=sign(log2FC))
)
```

Here, we will export a text file (containing the median of each transcript across inferential replicates) for additional plotting. First, select all features, use abind to combine multi-dimensional arrays, and then take the median:

```{r}
infReps <- assays(y)[ grep("infRep", assayNames(y)) ]
infArray <- abind::abind( as.list(infReps), along=3 )
dim(infArray)

infMed <- apply(infArray, 1:2, median)
data.class(infMed)

write.table(infMed, file = "DTE_InfMed.txt", sep = "\t")
```

Plot a histogram to show the distribution of pvalues:

```{r}
hist(mcols(y)$pvalue, col="grey", ylim= c(0,25000), main = "", xlab = "Pvalue",
     cex.axis = 0.9)
```

Add symbols (Gene IDs) to the *SummarizedExperiment* object. We will label genes on the MA plot in later steps. First load the *org.Mm.eg.db* package:

```{r}
suppressPackageStartupMessages(library(org.Mm.eg.db)) #Version 3.16.0
y <- addIds(y, "SYMBOL", gene=TRUE)
rowData(y)
```

Turn 'y' into a dataframe here, and save as a different variable. We do this to intersect with other lists when making plots, such as genes to highlight on the MA plot downstream. Check the *dim* to ensure that the size of the data frame has not changed:

```{r}
y_rowData <- as.data.frame(rowData(y))
dim(y_rowData)
```

Here we graph an MA plot. An MA plot shows log2FC vs the log10mean. Transcripts above M=0 are upregulated, while transcripts below are downregulated.

The transcripts that are farthest away from the y intercept (have the largest log10mean values) are the ones most affected by the treatment.

Set colors for the MA plot here: Transcripts that do not have a significant change in expression are light grey. Transcripts that are significantly affected (qvalue \< 0.05) by SD are dark grey.

```{r}
Significant_Color <- brewer.pal(9, "Greys")[8]
```

A subset of transcripts that we have chosen to highlight that also have significant changes in expression (qvalue \< 0.05) will be bright blue. Downstream, we will apply an abs(log2FC) threshold of 0.2 to these transcripts.

```{r}
Highlight_Color <- "dodgerblue" 
```

Read the list of genes that we have chosen to highlight here.

```{r}
Highlight_Genes <- readxl::read_excel("050123_Highlight_Genes_Subset.xlsx", 
                                      col_names = FALSE)
Highlight_Genes
```

Save 'Highlight_Genes' as a matrix titled 'Gene_Names':

```{r}
Gene_Names <- as.matrix(Highlight_Genes)
dim(Gene_Names)
```

Remove commas, quotes and trailing spaces:

```{r}
Gene_Names <- unlist(strsplit(Gene_Names,","))
Gene_Names <- noquote(Gene_Names)
Gene_Names <- trimws(Gene_Names)
```

We will intersect 'y_rowData' (the dataframe saved above) with the 'Gene_Names' matrix just saved using the *filter* function of *dplyr*:

```{r}
Data_Intersection <- dplyr::filter(y_rowData, SYMBOL %in% Gene_Names & qvalue < 0.05 & abs(log2FC) > 0.2)
dim(Data_Intersection)
```

Use *plotMASwish* to generate the MA plot. Include labels and points for the genes in 'Gene_Names' using the code below (Only transcripts that fit the qvalue and log2FC threshold will be labeled):

```{r}
plotMASwish(y, alpha=.05, xlim=c(0,6), ylim=c(-8,5),
            sigcolor = Significant_Color, cex.axis = 1.3, cex.lab = 1.3)

with(
  subset(mcols(y), SYMBOL %in% Gene_Names & qvalue < .05 & abs(log2FC) > 0.2),
  text(log10mean, log2FC, SYMBOL,
       col= Highlight_Color, pos=4, cex=0.8, font=2) 
)

points(Data_Intersection$log10mean, Data_Intersection$log2FC, pch= 0, col = Highlight_Color, cex = 1, lwd = 1)

points(Data_Intersection$log10mean, Data_Intersection$log2FC, pch= 20, col = Highlight_Color, cex = 0.2, lwd = 2)

```

Write text file using the code below. Select columns to include in the text file. These lists will then be annotated in Perl with the gene name and description.

```{r}
cols <- c("log10mean","log2FC","pvalue","qvalue", "SYMBOL")
y_filter <- mcols(y)
y_filter <- as.data.frame(y_filter)[,cols]
y_filter <- dplyr::filter(y_filter, qvalue < .05)
dim(y_filter)
write.table(y_filter, file = "DTE_Significant_k=4.txt")
```

# Differential Gene Expression Starts Here

Use the *summarizeToGene* function to proceed with analysis at the gene level. For the remainder of the analysis, 'gse' will be referred to as 'gy':

```{r}
gse <- summarizeToGene(se)
gy <- gse
```

I will do the same two group comparison between wild-type animals that was done at the transcript level. To quickly summarize this again, we will compare animals that were left undisturbed in their home cage (HC) for 5 hours and those who were sleep sleep deprived (SD) via gentle handling for 5 hours.

As above, the following analysis will compare the second group to the first group listed in *levels*, in this case, SD compared to HC controls.

```{r}
gy <- gy[,gy$condition %in% c("WTHC5", "WTSD5")]
gy$condition <- factor(gy$condition, levels=c("WTHC5", "WTSD5"))
```

## Proceed with Differential Expression Analysis

Scale inferential replicates to the mean sequencing depth. It is important to note that only the inferential replicates are scaled during this process, not the counts.

```{r}
gy <- scaleInfReps(gy)
```

Before filtering, we have 54,347 genes:

```{r}
dim(gy)
```

Filter here, keeping a minimum count of 10 transcripts across a minimum of 3 replicates. In this step, we are removing most of the rows with zero to very low counts before proceeding with the analysis.

```{r}
gy <- labelKeep(gy, minCount = 10, minN = 3) 
```

After filtering, we have 18,334 genes:

```{r}
gy <- gy[mcols(gy)$keep,]
dim(gy)
```

## Determining the Sources of Unwanted Variation with RUVs

During transcript expression analysis we already assembled a matrix that contains the groups. View it here to make sure 'groups' appears as expected.

```{r}
groups
```

We will use the same colors that were used for differential transcript expression, so this step does not need to be repeated. Again, HC controls will be light grey and SD samples will be dark grey.

Remove the version #s for each gene. It is helpful to note that the version #s correspond to the release version:

```{r}
rownames(gy) <- lapply(rownames(gy),  sub, pattern = "\\.\\d+$", 
                       replacement = "")
data.class(rownames(gy))
```

Extract the counts from the *SummarizedExperiment* object for normalization. Note that we will estimate the factors of unwanted variation on the counts, and then we will apply that across the inferential replicates further downstream:

```{r}
counts_gy <- as.matrix((assays(gy)[["counts"]]))
data.class(counts_gy)
```

For gene level analysis, we can use a set of positive controls from [@gerstner2016] to assess the performance of our pipeline downstream. Load the positive controls here:

```{r}
Gene_Positive_Controls <- read.table("Additional_File2_Positive_Controls.txt", header = TRUE)
dim(Gene_Positive_Controls)
```

In preparation for determining our positive control recovery, we will then want to intersect the expressed genes with the positive controls.

```{r}
Intersect_Positive_Expressed <- intersect(Gene_Positive_Controls[, 1], row.names(counts_gy))
length(Intersect_Positive_Expressed)
```

Upper Quartile (UQ) normalization is implemented here with the *EDASeq* package (same as with transcript expression analysis), and accounts for variation in sequencing depth:

As the counts were not scaled (only the inferential replicates), we need to account for sequencing depth prior to proceeding with *RUVs* normalization, which allows us to obtain a the estimated factors of unwanted variation (*gs\$W*):

```{r}
g_uq <- betweenLaneNormalization(counts_gy, which = "upper")
dim(g_uq)
```

Shorten the sample names (*colnames*) for figures:

```{r}
colnames(g_uq)
colnames(g_uq) <- c("HC5_1", "HC5_2", "HC5_3", "HC5_4", "HC5_5", "SD5_1", 
                    "SD5_2", "SD5_3", "SD5_4", "SD5_5")
```

Here we plot two quality control plots (RLE and PCA) following UQ normalization. Note that triangles are sleep deprived animals and circles are home cage control animals that were collected at the same circadian time:

```{r}
plotRLE(g_uq, col= colLib, outline = FALSE, las = 3, ylim = c(-1, 1), 
        ylab = "Relative Log Expression", cex.axis = 1.3, cex.lab = 1.3)
plotPCA(g_uq, labels = FALSE, col = colLib, cex = 1.5, cex.axis = 1.3, cex.lab = 1.3, 
        xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75), pch = rep(c(19,17), times = c(5,5)))
```

*RUVseq*: Here we will estimate a matrix that contains estimated factors of unwanted factors after UQ normalization.

*RUVs* uses technical replicates or negative controls samples. We do have negative control genes from [@gerstner2016] for differential gene analysis, and will load them here. For additional information regarding *RUVs*, see [@risso2014].

```{r}
Gene_Negative_Controls <- read.table("Final_Negative_Controls_v25.txt", header = TRUE)
dim(Gene_Negative_Controls)
```

After loading negative controls, intersect with the counts, to only include the negative controls that are expressed in the dataset:

```{r}
Negative <- intersect(Gene_Negative_Controls[, 1], rownames(counts_gy))
length(Negative)
```

After viewing the results of the differential expression analysis with several *k*s, we have set the numbers of factors of unwanted variation (*k*) = 4, which optimizes the number of differentially expressed genes and transcripts detected, and the positive control recovery (at the gene level) without removing signal.

```{r}
gs <- RUVs(x = g_uq, cIdx = Negative, scIdx = groups, k = 4)
```

Plot the RLE and PCA plots (quality control figures) again following *RUVs* normalization.

```{r}
plotRLE(gs$normalizedCounts, col= colLib, outline = FALSE, las = 3, 
        ylim = c(-1, 1), ylab = "Relative Log Expression", cex.axis = 1.2, 
        cex.lab = 1.2)
plotPCA(gs$normalizedCounts, labels = FALSE, col = colLib, cex = 1.3, 
        cex.axis = 1.2, cex.lab = 1.2, xlim = c(-0.75, 0.75), ylim = c(-0.75, 0.75), 
        pch = rep(c(19,17), times = c(5,5)))
```

Downstream we will use *gs\$W* (from *RUVs*) as the estimation of batch factors.

'infRepIdx' lists the inferential replicates (30):

```{r}
infRepIdx <- grep("infRep",assayNames(gy),value=TRUE)
```

Save the number of inferential replicates as 'nreps_gy':

```{r}
nreps_gy <- length(infRepIdx)
nreps_gy
```

Account for continuous variables with *removeBatchEffect* from *limma*. Our samples do not fall into discrete clusters so we will use the following approach recommended by the *fishpond* developers.

In short, this is done by directly scaling the estimated counts across inferential replicates.

First inferential replicates are logged as *limma* requires log-expression values for a series of samples. This is done with the *assay* function from the *SummarizedExperiment* package.

First, we load the *limma* package:

```{r}
model_matrix_gy <- model.matrix(~condition, colData(gy))
pc <- .1 # This is added to avoid negative InfReps
for (k in seq_len(nreps_gy)) {
  logInfRep <- log(assay(gy, infRepIdx[k]) + pc)
  logInfRep <- limma::removeBatchEffect(
    logInfRep,
    covariates=gs$W,
    design=model_matrix_gy)
  assay(gy, infRepIdx[k]) <- exp(logInfRep)
}
```

## Run Swish and Plot Additional Figures

The Swish method is described in [@zhu2019]. Note that the *set.seed* function allows for the reproducibility of exact results in the future.

```{r}
set.seed(1)
gy <- swish(gy, x="condition")
```

View a table of differently expressed genes with a qvalue \< 0.05:

```{r}
table(mcols(gy)$qvalue < .05)
```

You can also view a table that shows which genes are up regulated (1) in response to sleep deprivation and which are downregulated (-1):

```{r}
with(mcols(gy),
     table(sig=qvalue < .05, sign.lfc=sign(log2FC))
)
```

Here, we can assess our positive control recovery:

```{r}
significant_qvalue_gy <- (mcols(gy)[mcols(gy)$qvalue < .05,])
dim(significant_qvalue_gy)

Positive_Controls_Recovery <- length(intersect(row.names(significant_qvalue_gy), Intersect_Positive_Expressed))

(length(intersect(row.names(significant_qvalue_gy), Intersect_Positive_Expressed))/length(Intersect_Positive_Expressed)) *100
```

Plot a histogram to show the distribution of pvalues:

```{r}
hist(mcols(gy)$pvalue, col="grey", ylim= c(0,10000), main = "", xlab = "Pvalue",
     cex.axis = 0.9)
```

Add symbols (Gene IDs) to the *SummarizedExperiment* object, just like we did with transcript expression analysis.

```{r}
gy <- addIds(gy, "SYMBOL", gene=TRUE)
rowData(gy)
```

Turn 'gy' into a dataframe here, and save as a different variable. We do this to intersect with other lists when making plots, such as genes to highlight on the MA plot downstream. Check the *dim* to ensure that it has not changed:

```{r}
gy_rowData <- as.data.frame(rowData(gy))
dim(gy_rowData)
```

Here we graph an MA plot. An MA plot shows log2FC vs the log10mean. Genes above M=0 are upregulated, while genes below are downregulated.

The genes that are farthest away from the y intercept (have the largest log10mean values) are the ones most affected by the treatment.

We previously set colors for the MA plot, so we will only set one additional color here: the positive controls.

A quick summary of the other colors: Genes that do not have a significant change in expression are light grey. Genes that are significantly affected (qvalue \< 0.05) by SD are dark grey. Genes that we have chosen to highlight that are also significant are bright blue.

```{r}
Positive_Color <- brewer.pal(9, "Set1")[1]
```

We will intersect 'gy_rowData' (the dataframe saved above) with the 'Gene_Names' matrix from transcript expression analysis:

```{r}
Data_Intersection <- dplyr::filter(gy_rowData, SYMBOL %in% Gene_Names & qvalue < 0.05 & abs(log2FC) > 0.2)
dim(Data_Intersection)
```

Use *plotMASwish* to generate the MA plot. Include labels and points for the genes in 'Gene_Names' and points for positive controls using the code below (Only genes that fit the qvalue and log2FC threshold will be labeled):

```{r}
plotMASwish(gy, alpha=.05, xlim=c(.5,5.5), ylim=c(-7,5), sigcolor = Significant_Color,
            cex.axis = 1.3, cex.lab = 1.3)

with(
  subset(mcols(gy), SYMBOL %in% Gene_Names & qvalue < .05 & abs(log2FC) > 0.2),
  text(log10mean, log2FC, SYMBOL,
       col= Highlight_Color, pos=4, cex=0.8, font=2) 
)

Positive_Controls_Signifcant <- intersect(row.names(significant_qvalue_gy), Intersect_Positive_Expressed)
Positive_Controls_Plot <- mcols(gy)[ rownames(mcols(gy)) %in% Positive_Controls_Signifcant, ]
dim(Positive_Controls_Plot)
points(x= Positive_Controls_Plot$log10mean, y = Positive_Controls_Plot$log2FC, pch = 20, col = Positive_Color, cex = .2, lwd = 2)

points(Data_Intersection$log10mean, Data_Intersection$log2FC, pch= 0, col = Highlight_Color, cex = 1, lwd = 1)

points(Data_Intersection$log10mean, Data_Intersection$log2FC, pch= 20, col = Highlight_Color, cex = 0.2, lwd = 2)

```

Write text file using the code below. Select columns to include in the text file. These lists will then be annotated in Perl with the gene name and description.

```{r}
cols <- c("log10mean","log2FC","pvalue","qvalue")
gy_filter <- mcols(gy)
gy_filter <- as.data.frame(gy_filter)[,cols]
gy_filter <- dplyr::filter(gy_filter, qvalue < .05)
dim(gy_filter)
write.table(gy_filter, file = "DGE_Significant_k=4.txt")
```

## Session Info

```{r}
sessionInfo()
```

## References
